File: frontend/src/types/auditLog.ts
export interface AuditLog {
  id: number;

  user_id: number;

  user_name: string;

  user_email: string;

  action: string;

  resource_type: string | null;

  resource_id: string | null;

  description: string | null;

  entity_name: string | null;

  ip_address: string | null;

  user_agent: string | null;

  status: string;

  created_at: string;
}


// ==============================
// Audit Log Detail
// ==============================

export interface AuditLogDetail
  extends AuditLog {

  before_values:
    | Record<string, unknown>
    | null;

  after_values:
    | Record<string, unknown>
    | null;
}


// ==============================
// Audit Log Pagination Response
// ==============================
export interface AuditLogPage {
  items: AuditLog[];

  total: number;

  page: number;

  limit: number;

  total_pages: number;
}


// =================
// Audit Log Filters
// ==================
export interface AuditLogFilters {
  page: number;

  limit: number;

  search?: string;

  user_id?: number;

  action?: string;

  resource_type?: string;

  status?: string;

  start_date?: string;

  end_date?: string;

  sort:
    | "newest"
    | "oldest";
}

File: backend/app/schemas/audit_log.py
from datetime import datetime
from typing import Any, Optional
from pydantic import BaseModel, ConfigDict, Field


class AuditLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    user_id: int
    user_name: str
    user_email: str
    action: str
    resource_type: Optional[str] = None
    resource_id: Optional[str] = None
    description: Optional[str] = None
    entity_name: Optional[str] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    status: str
    created_at: datetime


class AuditLogDetail(AuditLogOut):
    before_values: Optional[dict[str, Any]] = None
    after_values: Optional[dict[str, Any]] = None


class AuditLogPage(BaseModel):
    items: list[AuditLogOut]
    total: int
    page: int = Field(ge=1)
    limit: int = Field(ge=1, le=100)
    total_pages: int
